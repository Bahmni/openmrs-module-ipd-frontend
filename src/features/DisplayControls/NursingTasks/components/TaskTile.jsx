import React from "react";
import PropTypes from "prop-types";
import SVGIcon from "../../../SVGIcon/SVGIcon";
import Clock from "../../../../icons/clock.svg";
import {
  getTime,
  getRelevantTaskStatus,
  iconType,
} from "../utils/TaskTileUtils";
import { Tag, TooltipDefinition } from "carbon-components-react";
import "../styles/TaskTile.scss";
import { FormattedMessage } from "react-intl";

export default function TaskTile(props) {
  const { medicationNursingTask } = props;
  const newMedicationNursingTask = medicationNursingTask[0];
  let isGroupedTask, taskCount;
  if (medicationNursingTask.length > 1) {
    isGroupedTask = true;
    taskCount = medicationNursingTask.length - 1;
  }

  const {
    drugName,
    dosage,
    doseType,
    drugRoute,
    startTime,
    startTimeInEpochSeconds,
    stopTime,
    isDisabled,
    administeredTimeInEpochSeconds,
  } = newMedicationNursingTask;

  const isRelevantTask = getRelevantTaskStatus(startTimeInEpochSeconds);

  const drugNameText = (
    <div
      className="drug-title"
      style={{
        color: stopTime ? "#da1e28" : isRelevantTask ? "#393939" : "#525252",
        fontWeight: isRelevantTask ? 500 : 400,
      }}
    >
      {drugName}
    </div>
  );
  return (
    <div className="tile-parent-container">
      <div
        className={`nursing-tasks-tile ${stopTime && "no-hover"}
        ${isRelevantTask && !stopTime && "relevant-task-tile"} ${
          isDisabled ? "disabled-tile no-hover" : ""
        }`}
      >
        <div className="tile-content">
          <div className={`tile-title ${stopTime && "red-text"}`}>
            <div>
              <div
                className="nursing-task-icon-container"
                data-testid={iconType(
                  administeredTimeInEpochSeconds,
                  startTimeInEpochSeconds
                )}
              >
                <SVGIcon
                  iconType={iconType(
                    administeredTimeInEpochSeconds,
                    startTimeInEpochSeconds
                  )}
                />
              </div>
              <TooltipDefinition
                tooltipText={drugName}
                className={stopTime ? "cursor-not-allowed" : "cursor-pointer"}
              >
                {drugNameText}
              </TooltipDefinition>
            </div>
            {stopTime && (
              <Tag className={"red-tag"}>
                <FormattedMessage id={"STOPPED"} defaultMessage={"Stopped"} />
              </Tag>
            )}
          </div>
          <div
            className="tile-content-subtext"
            style={{ color: isRelevantTask ? "#393939" : "#525252" }}
          >
            <span>{dosage}</span>
            {doseType && <span>&nbsp;-&nbsp;{doseType}</span>}
            <span>&nbsp;-&nbsp;{drugRoute}</span>
          </div>
          <div className="tile-content-subtext">
            <Clock />
            <div>
              &nbsp;{getTime(administeredTimeInEpochSeconds, startTime)}
            </div>
          </div>
          {isGroupedTask && <div className="more-info">({taskCount} more)</div>}
        </div>
      </div>
      {isGroupedTask && (
        <div
          className={`nursing-tasks-tile stacked-tile ${
            isRelevantTask && "relevant-task-tile"
          }`}
        ></div>
      )}
    </div>
  );
}
TaskTile.propTypes = {
  medicationNursingTask: PropTypes.array.isRequired,
};
