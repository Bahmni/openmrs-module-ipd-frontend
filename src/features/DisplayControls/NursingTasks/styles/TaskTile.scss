.nursing-tasks-tile {
  border: 1px solid #007d79;
  background-color: #edf8e6;
  width: 323px;
  height: 90px;
  position: relative;
  cursor: pointer;

  &:hover {
    background-color: #d2e1c9;
  }
}

.relevant-task-tile {
  background-color: #e3fed1;

  &:hover {
    background-color: #c5eea9;
  }
}

.tile-parent-container {
  margin-bottom: 20px;
  margin-right: 20px;
  height: 90px;
}

.stacked-tile {
  position: relative;
  top: -80px;
  left: 8px;
  z-index: -1;
}

.tile-content {
  padding: 10px;
}

.tile-title {
  flex-direction: row;
  display: flex;

  .bx--tooltip__trigger.bx--tooltip__trigger--definition.bx--tooltip--bottom {
    cursor: pointer;
  }
}

.drug-title {
  white-space: nowrap;
  overflow: hidden;
  text-wrap: wrap;
  color: #393939;
  font-style: normal;
  padding-left: 4px;
  font-size: 14px;
  line-height: normal;
  letter-spacing: 0.14px;
  text-align: left;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 2;
}

.tile-content-subtext {
  margin-top: 4px;
  font-size: 12px;
  font-weight: 400;
  color: #393939;
  letter-spacing: 0.12px;
  align-items: center;
  display: flex;
}

.nursing-task-icon-container {
  margin-top: 2px;
}

.more-info {
  margin-top: 4px;
  font-size: 12px;
  font-weight: 400;
  color: #393939;
  letter-spacing: 0.12px;
  position: absolute;
  bottom: 4px;
  right: 8px;
}

.bx--tooltip__trigger.bx--tooltip__trigger--definition {
  border-bottom: 0;
}

.bx--tooltip__trigger.bx--tooltip__trigger--definition.bx--tooltip--bottom.bx--tooltip--align-start.bx--tooltip--a11y
  + .bx--assistive-text {
  color: #fff;
  font-size: 12px;
  font-style: normal;
  font-weight: 400;
  line-height: 15px;
  letter-spacing: 0.16px;
}
