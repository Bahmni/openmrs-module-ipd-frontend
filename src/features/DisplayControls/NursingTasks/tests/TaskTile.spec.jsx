import React from "react";
import { render, waitFor } from "@testing-library/react";
import MockDate from "mockdate";

import TaskTile from "../components/TaskTile";
import {
  mockCompletedTaskTileData,
  mockPendingTaskTileData,
  mockTaskTileDataForGroupedTask,
} from "./NursingTasksUtilsMockData";

describe("TaskTile", () => {
  beforeEach(() => {
    MockDate.set("2023-08-11");
  });

  afterEach(() => {
    MockDate.reset();
  });
  it("should match the snapshot for non grouped task", () => {
    const { asFragment } = render(
      <TaskTile medicationNursingTask={mockCompletedTaskTileData} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should match the snapshot for grouped task", async () => {
    const { asFragment, getByText } = render(
      <TaskTile medicationNursingTask={mockTaskTileDataForGroupedTask} />
    );

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
      expect(getByText("(1 more)")).toBeTruthy();
    });
  });

  it("should disable tile when task is completed", () => {
    const { container, getByTestId } = render(
      <TaskTile medicationNursingTask={mockCompletedTaskTileData} />
    );

    const completedTile = container.querySelector(".disabled-tile");
    const administeredIcon = getByTestId("Administered");

    expect(completedTile).toBeTruthy();
    expect(administeredIcon).toBeTruthy();
  });

  it("should render pending icon when task is not administered", () => {
    const { getByTestId } = render(
      <TaskTile medicationNursingTask={mockPendingTaskTileData} />
    );

    const pendingIcon = getByTestId("Pending");

    expect(pendingIcon).toBeTruthy();
  });
});
