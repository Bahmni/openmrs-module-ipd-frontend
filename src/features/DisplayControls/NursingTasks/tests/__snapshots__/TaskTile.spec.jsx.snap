// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskTile should match the snapshot for grouped task 1`] = `
<DocumentFragment>
  <div
    class="tile-parent-container"
  >
    <div
      class="nursing-tasks-tile false 
        undefined"
    >
      <div
        class="tile-content"
      >
        <div
          class="tile-title undefined"
        >
          <div>
            <div
              class="nursing-task-icon-container"
              data-testid="Pending"
            >
              <div>
                <test-file-stub />
              </div>
            </div>
            <span
              class="bx--tooltip--definition bx--tooltip--a11y cursor-pointer"
            >
              <button
                aria-describedby="definition-tooltip-2"
                class="bx--tooltip__trigger bx--tooltip--a11y bx--tooltip__trigger--definition bx--tooltip--bottom bx--tooltip--align-start"
                type="button"
              >
                <div
                  class="drug-title"
                  style="color: rgb(82, 82, 82); font-weight: 400;"
                />
              </button>
              <span
                class="bx--assistive-text"
                id="definition-tooltip-2"
                role="tooltip"
              />
            </span>
          </div>
        </div>
        <div
          class="tile-name-cell"
        >
          <div
            class="bx--tag bx--tag--blue"
            id="tag-2"
          >
            <span
              title="Rx"
            >
              Rx
            </span>
          </div>
        </div>
        <div
          class="tile-content-subtext"
          style="color: rgb(82, 82, 82);"
        >
          <span />
          <span>
             - 
          </span>
        </div>
        <div
          class="tile-content-subtext"
        >
          <test-file-stub />
          <div>
             
          </div>
        </div>
        <div
          class="more-info"
        >
          (1 more)
        </div>
      </div>
    </div>
    <div
      class="nursing-tasks-tile stacked-tile false"
    />
  </div>
</DocumentFragment>
`;

exports[`TaskTile should match the snapshot for non grouped task 1`] = `
<DocumentFragment>
  <div
    class="tile-parent-container"
  >
    <div
      class="nursing-tasks-tile false 
        disabled-tile"
    >
      <div
        class="tile-content"
      >
        <div
          class="tile-title undefined"
        >
          <div>
            <div
              class="nursing-task-icon-container"
              data-testid="Administered"
            >
              <div>
                <test-file-stub />
              </div>
            </div>
            <span
              class="bx--tooltip--definition bx--tooltip--a11y cursor-not-allowed"
            >
              <button
                aria-describedby="definition-tooltip-1"
                class="bx--tooltip__trigger bx--tooltip--a11y bx--tooltip__trigger--definition bx--tooltip--bottom bx--tooltip--align-start"
                type="button"
              >
                <div
                  class="drug-title"
                  style="color: rgb(82, 82, 82); font-weight: 400;"
                >
                  Halothane/Fluothane 250 mL Inhalation Anesthetic Liquid
                </div>
              </button>
              <span
                class="bx--assistive-text"
                id="definition-tooltip-1"
                role="tooltip"
              >
                Halothane/Fluothane 250 mL Inhalation Anesthetic Liquid
              </span>
            </span>
          </div>
        </div>
        <div
          class="tile-name-cell"
        >
          <div
            class="bx--tag bx--tag--blue"
            id="tag-1"
          >
            <span
              title="Rx"
            >
              Rx
            </span>
          </div>
        </div>
        <div
          class="tile-content-subtext"
          style="color: rgb(82, 82, 82);"
        >
          <span>
            1ml
          </span>
          <span>
             - Oral
          </span>
        </div>
        <div
          class="tile-content-subtext"
        >
          <test-file-stub />
          <div>
             16:00 - 17:12 (actual)
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
